api/src/API/Controllers/RaceResultsController.cs

```diff
@@ -51,12 +51,23 @@ public async Task<IActionResult> GetAllRaceResults(string orgId)
             var racesNeeded = raceResults.Select(result => result.RaceId).ToHashSet();
             var racesInResponse = await raceContainer.GetManyAsDictAsync(it => it.Where(race => racesNeeded.Contains(race.Id)));
 
-            var result = raceResults.Select(raceResult => new RaceResultResponse()
+            var result = new List<RaceResultResponse>();
+            foreach (var raceResult in raceResults)
             {
-                RaceResult = raceResult,
-                Member = members[raceResult.MemberId],
-                Race = racesInResponse[raceResult.RaceId],
-            });
+                try
+                {
+                    result.Add(new RaceResultResponse()
+                    {
+                        RaceResult = raceResult,
+                        Member = members[raceResult.MemberId],
+                        Race = racesInResponse[raceResult.RaceId],
+                    });
+                } catch (KeyNotFoundException)
```

Comment:
I'd prefer if we don't do this. If our invariants for the representation of the database data are violated, this would result in a silent failure. We should make sure that we never run into this situation, and if we do, throw an error so that we can look into the issue

Reply:
That's true. However since we're not using SQL we don't really have invariants since there's nothing enforcing foreign key constraints. For example if we decide to delete a Race, should we also delete every RaceResult that was for it?

Something we can do is make the response have nullable instances of Member and Race, and return null for those fields if they don't exist. Thoughts?

Reply:
By "don't really have invariants" I mean on a DB level. We have business logic invariants only

Reply:
Yes, I'm talking about business logic invariants. I actually do think we should either never delete races or delete/update every race result that was for it.

What do you mean return null? Are you talking about returning null from the db call? from the api call?

Reply:
I don't have much experience in this field, but I would think it's best practice to avoid enforcing business logic invariants in the DB.

The problem here is also that we're being forced to _enforce database invariants at the API level_. I think we should avoid this. It makes sense to not bring down an entire route just because a document doesn't exist. For example, imagine a social media site where a Post document contains `CommentId[]`. I don't think it would make sense to bring down `/post/{postId}/comments` just because one of the ids in `CommentIds` doesn't exist. Since the database isn't enforcing this one-to-many relationship, anything could have happened to cause it - maybe the DB isn't done replicating that comment yet; perhaps the comment ID was lost in POST or just isn't done yet (which, considering the update to the Post document and creation of the Comment document are not atomic, can happen).

By return null, I mean `RaceResultResponse.Race` and `RaceResultResponse.Member` can be null if a document with `RaceResultResponse.RaceId` or `RaceResultResponse.MemberId` does not exist. The frontend will display this as some type of error for the submission in question.

Reply:
Yeah, I guess this is just a pitfall of the layers of abstraction chosen. The controller here is assuming the responsibility of interacting with multiple database containers at the same time. The enforcement I'm referring to is that of the implied spec of this controller method (I'll make a work item to write out the specs).

I'm not suggesting we enforce database invariants at the API level. I'm suggesting we enforce API specifications at the API level. For example, we might decide that deleting a race also deletes the associated race results which the method specification can guarantee an atomic operation.

Ultimately, I agree with you. we shouldn't bring down the whole endpoint for a single inconsistency but I think we should actively do something about it. I'll come up with some ideas and we can discuss them together.